### PR TITLE
Adds an ideal ROT functionality to mpz-ot-core

### PR DESCRIPTION
**Pull request by th4s**

### CODE REVIEW

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    /// * `receiver_msg` - The message that the ideal ROT sends to the receiver.
+    pub fn check(&self, sender_msg: RotMsgForSender, receiver_msg: RotMsgForReceiver) -> bool {
+        let RotMsgForSender { qs } = sender_msg;
+        let RotMsgForReceiver { rs, ts } = receiver_msg;
+
+        qs.into_iter()
+            .zip(ts)
+            .zip(rs)
+            .all(|((q, t), r)| if r { q[1] == t } else { q[0] == t })
+    }
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_rot.rs]: This is just a unit test and shouldn't be part of the API

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+/// The message that receiver receives from the ROT functionality.
+pub struct RotMsgForReceiver {
+    /// The random bits that receiver receives from the ROT functionality.
+    pub rs: Vec<bool>,
+    /// The chosen blocks that receiver receives from the ROT functionality.
+    pub ts: Vec<Block>,
+}
+#[allow(missing_docs)]
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_rot.rs]: Please fix spacing between struct defs and document this struct

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    counter: usize,
+    prg: Prg,
+}
+
+impl IdealROT {
+    /// Initiate the functionality
+    pub fn new() -> Self {
+        let prg = Prg::new();
+        IdealROT { counter: 0, prg }
+    }
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_rot.rs]: We can implement `Default` as well.

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+/// The message that receiver receives from the ROT functionality.
+pub struct RotMsgForReceiver {
+    /// The random bits that receiver receives from the ROT functionality.
+    pub rs: Vec<bool>,
+    /// The chosen blocks that receiver receives from the ROT functionality.
+    pub ts: Vec<Block>,
+}
+#[allow(missing_docs)]
+pub struct IdealROT {
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_rot.rs]:
```suggestion
#[derive(Debug)]
pub struct IdealROT {
```

**Review by sinui0 — CHANGES_REQUESTED** [review]
lgtm other than a few things

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs (hunk 3 above)
Comment (th4s) [reply ot/mpz-ot-core/src/ideal/ideal_rot.rs]: It is already implemented.

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs (hunk 4 above)
Comment (th4s) [reply ot/mpz-ot-core/src/ideal/ideal_rot.rs]: Unfortunately this does not work, because `Prg` does not implement `Debug`.

Comment (sinui0) [reply ot/mpz-ot-core/src/ideal/ideal_rot.rs]: Pls add `Debug` to `Prg` (probably opaque impl to hide the key)

File: ot/mpz-ot-core/src/ideal/ideal_rot.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    #[cfg(test)]
+    fn check(&self, sender_msg: RotMsgForSender, receiver_msg: RotMsgForReceiver) -> bool {
+        let RotMsgForSender { qs } = sender_msg;
+        let RotMsgForReceiver { rs, ts } = receiver_msg;
+
+        qs.into_iter()
+            .zip(ts)
+            .zip(rs)
+            .all(|((q, t), r)| if r { q[1] == t } else { q[0] == t })
+    }
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_rot.rs]: This doesn't belong here, it just adds noise to the impl block. This function should just be inlined in the unit test

**Review by sinui0 — APPROVED** [review]

**Review by themighty1 — APPROVED** [review]

**Comment by th4s** [comment]
I will wait for the IO refactor before merging this.

**Comment by sinui0** [comment]
lgtm

**Pull request merged**